### PR TITLE
Dropdown: implement adaptive Dropdown with SheetMobile for mobile devices

### PR DESCRIPTION
**Pull request by AlbertCarreras**
Dropdown: implement adaptive Dropdown with SheetMobile for mobile devices

Brave Browser - Dropdown - Gestalt 2023-04-27 at 11 58 53 PM

### CODE REVIEW

File: packages/gestalt/src/Dropdown.js [hunk 1]
```diff
@@ -122,10 +125,18 @@ type Props = {|
...
+   */
+  mobileOnAnimationEnd?: ({| animationState: 'in' | 'out' |}) => void,
   /**
    * Callback fired when the menu is closed.
    */
   onDismiss: () => void,
+  /**
+   * Dropdown can adapt to mobile devices to [SheetMobile](https://gestalt.pinterest.systems/web/sheetmobile). Mobile adaptation is not enabled by default. Set to 'sheetMobile' to enable SheetMobile in mobile devices. See the [mobile variant](https://gestalt.pinterest.systems/web/dropdown#mobile) to learn more.
+   */
+  mobileAdaptation?: 'default' | 'sheetMobile',
```
Comment (dangerismycat) [inline packages/gestalt/src/Dropdown.js:139]: Is the plan to eventually make `sheetMobile` the default? If so, we may want to make this something like `disableMobileUI` and codemod all current usages to include the prop, such that any new usages won't have it — therefore encouraging new Dropdown usages to embrace the adaptive UI.

File: packages/gestalt/src/Dropdown.js [hunk 2]
```diff
@@ -149,8 +160,13 @@ export default function Dropdown({
...
   zIndex,
   maxHeight,
+  mobileOnAnimationEnd,
+  mobileAdaptation = 'default',
 }: Props): Node {
-  const [hoveredItem, setHoveredItem] = useState<number>(0);
+  const deviceType = useDeviceType();
+  const isMobile = deviceType === 'mobile';
+
+  const [hoveredItem, setHoveredItem] = useState<?number>(isMobile ? undefined : 0);
```
Comment (dangerismycat) [inline packages/gestalt/src/Dropdown.js:169]: nit: this is the index of the hovered item, yes? If so, consider `hoveredItemIndex` to be more specific. We could also use an id or the item reference itself, so clarifying helps readability.

File: packages/gestalt/src/DropdownLink.js [hunk 3]
```diff
@@ -41,6 +41,7 @@ type Props = {|
   onClick?: ({|
     event: SyntheticMouseEvent<HTMLAnchorElement> | SyntheticKeyboardEvent<HTMLAnchorElement>,
     dangerouslyDisableOnNavigation: () => void,
+    mobileOnDismissStart: () => void,
```
Comment (dangerismycat) [inline packages/gestalt/src/DropdownLink.js:44]: Can you add documentation about this new field?

File: packages/gestalt/src/SheetMobile.js [hunk 4]
```diff
@@ -104,6 +105,10 @@ type Props = {|
    * Sets the SheetMobile's height. See the [size variant](https://gestalt.pinterest.systems/web/sheetmobile#Size) for more info.
    */
   size?: 'default' | 'full' | 'auto',
+  /**
+   * An object representing the zIndex value of the Dropdown menu. Learn more about [zIndex classes](https://gestalt.pinterest.systems/web/zindex_classes)
```
Comment (dangerismycat) [inline packages/gestalt/src/SheetMobile.js:109]: This should reference SheetMobile and not Dropdown, yes?

Comment (AlbertCarreras) [reply packages/gestalt/src/SheetMobile.js:109]: DONE

File: packages/gestalt/src/DropdownLink.js (hunk 3 above)
Comment (AlbertCarreras) [reply packages/gestalt/src/DropdownLink.js:44]: DONE

File: packages/gestalt/src/Dropdown.js (hunk 2 above)
Comment (AlbertCarreras) [reply packages/gestalt/src/Dropdown.js:169]: DONE

File: packages/gestalt/src/Dropdown.js (hunk 1 above)
Comment (AlbertCarreras) [reply packages/gestalt/src/Dropdown.js]: DONE

Keeping disabled, I'll run codemod next and invert default as a follow up. less blocking PR

File: packages/gestalt/src/Dropdown.js [hunk 5]
```diff
@@ -149,8 +160,13 @@ export default function Dropdown({
   onDismiss,
   zIndex,
   maxHeight,
+  mobileOnAnimationEnd,
+  disableMobileUI = true,
```
Comment (dangerismycat) [inline packages/gestalt/src/Dropdown.js:164]: This is an anti-pattern, leading to a confusing DX. For example, which of these yields a disabled mobile UI?
```js
<Dropdown />
<Dropdown disableMobileUI />
<Dropdown disableMobileUI={false} />
```
The first two both do, but how is a developer to know that the first example would disable the mobile UI?

We also plan to make this behavior the default in the future, requiring another breaking change in the future.

We can avoid that future breaking change and encourage all new usages of Dropdown to use the adaptive UI by setting this to our target default of `disableMobileUI = false`, and codemodding all current Dropdown usages.

**Review by dangerismycat — CHANGES_REQUESTED** [review]

**Comment by rlingineni** [comment]
I can also take a look at this tomorrow, since it seems like a big PR

File: packages/gestalt/src/Dropdown.js (hunk 5 above)
Comment (AlbertCarreras) [reply packages/gestalt/src/Dropdown.js:164]: Ryan, 

In my comment above:

Screenshot 2023-05-03 at 9 27 29 PM

I'm going to run the codemod, independently from the PR, set all current instances to true, then set default to false

@dangerismycat

**Review by dangerismycat — APPROVED** [review]

**Pull request merged**